### PR TITLE
Application Factory Method

### PR DESCRIPTION
**Pull request by softbobo**
This PR moves app creation to using a [factory method](https://flask.palletsprojects.com/en/1.1.x/tutorial/factory/). This makes sense for our testing framework and might be beneficial when needing to create several app instances down the road.

Packed this into a small PR of its own, so we are able to merge quickly and can continue developing the unit tests.

### CODE REVIEW

**Comment by softbobo** [comment]
refactored app creation a little bit, so tests pass now. still a long way to go

**Comment by softbobo** [comment]
Refactored errors subsystem for better maintainability. App factory creation now seems to work as expected

File: api/src/errors/routes.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from flask import make_response
+
+from src.errors import blueprint
+from src import metrics
+
+@blueprint.app_errorhandler(404)
```
Comment (obitech) [inline api/src/errors/routes.py]: What's the reason for this change?

Comment (softbobo) [reply api/src/errors/routes.py]: that we use a blueprint instead of decorating the app instance? there are several reasons:
- we prevent circular imports at the current app architecture
- we do not have get duplicated timeseries with prometheus
- it fits with the modular architecture of our app

File: api/src/__init__.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+   
+    metrics.init_app(app)
+
+    api.init_app(blueprint)   
+    api.add_resource(Healthz, "/healthz")
+
+    from src.errors import blueprint as errors_bp
+    app.register_blueprint(errors_bp) 
+
+    app.register_blueprint(blueprint)
```
[the remark this answers is not in the document]
Comment (softbobo) [reply api/src/__init__.py]: the second blueprint is legacy and is the `v1` prefix for the API. i was not entirely sure how to integrate it, so i just kept it.

File: api/Makefile [hunk 3]
```diff
@@ -11,6 +11,7 @@ help:
 .PHONY: test
 ## test: run unit tests
 test:
+	export CONFFILE=testconf.py; \
 	cd src/ && $(PY) -m pytest -s -x
```
[the remark this answers is not in the document]
Comment (softbobo) [reply api/Makefile]: not entirely sure what you mean here. as i can see it

[the remark this answers is not in the document]
Comment (softbobo) [reply api/Makefile]: > Looks good so far but a couple things are missing:
> 
>     * Running `make run/prod` locally will throw exceptions. Do we need to add a `CONFFILE` here as well? Please fix.

Will be fixed

>     * `devconf.py` and `testconf.py` are exactly the same, do we need both right now?

`testconf.py` will be temporarily deleted and recreated when we are clear on whether we need a special testing configuration and how it shall look like
 
>     * Documentation missing on new blueprints & why they are used

Will be fixed for the newly introduced one. Cannot add documentation on the legacy one as I am not exactly sure on why we use it anyways, see below.
 
>     * Switching to a new pattern is a big step but it's not clear how the changes are incorporated. Please squash all commits into a single one and provide a meaningful commit message explaining the changes made (think of it like a mini blog post for future contributors).

Will be fixed.
> 
> Generally I'm in favour of the factory pattern.

**Comment by softbobo** [comment]
squashed all into a single commit and integrated suggested changes as good as possible

File: api/Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (obitech) [reply api/Makefile]: At the top of the `Makefile`, we need to define the variable `CONFFILE`, similar to how `MULTIPROC_DIR` is defined:

```
MULTIPROC_DIR ?= $(shell pwd)/.multiproc_dir
```

Like:

```
SH := /bin/sh
PY := $(shell which python3)
MULTIPROC_DIR ?= $(shell pwd)/.multiproc_dir
CONFFILE = devconf.py
```

Then at the target we pass the make variable to the environment variable:

```
.PHONY: test
## test: run unit tests
test:
    CONFFILE=$(CONFFILE) cd src/ && $(PY) -m pytest -s -x
```

Like that we can set the environment variable when we invoke the make target:

```
make test CONFFILE=newfile.py
```

[the remark this answers is not in the document]
Comment (softbobo) [reply api/Makefile]: done

**Comment by softbobo** [comment]
Squashed everything. Should be nice, clean and revertable now?

**Comment by obitech** [comment]
Looks good! Next time please pick a good commit summary as the first line, then some description below. Right now it's all in a single line which gets truncated. Also please pick present tense for the commit message. There is a link in our `CONTRIBUTING.md` with more information on good commit messages: https://chris.beams.io/posts/git-commit/
